### PR TITLE
feat(theme): follow system theme by default

### LINKED ISSUE #182: feat: follow system theme
just sync it with the os 🥴

### PR DESCRIPTION
**Pull request by Rudxain**
closes #182

ℹ️ https://github.com/Universal-Debloater-Alliance/universal-android-debloater-next-generation/issues/182#issuecomment-2208347197

### CODE REVIEW

**Comment by Rudxain** [comment]
I'm having problems selecting `Auto`. For some reason, the GUI defaults to `Dark`. The logs don't give any useful info, it's similar to setting. The persistent `config.toml` file does store the `Auto` selection, but it doesn't apply on start. I suspect this bug is related to Serde

**Review by lavafroth — DISMISSED** [review]

File: src/core/theme.rs [hunk 1]
```diff
@@ -109,6 +127,8 @@ impl std::fmt::Display for Theme {
                 Self::Dark => "Dark",
                 Self::Light => "Light",
                 Self::Lupin => "Lupin",
+                // should it be "Follow System"?
+                Self::Auto => "Auto (sync on startup)",
```
Comment (Rudxain) [inline src/core/theme.rs:131]: Since I can't test `Auto`, I have no idea if it actually syncs on start, or if there are other events that update it

**Comment by Frigyes06** [comment]
> I'm having problems selecting `Auto`. For some reason, the GUI defaults to `Dark` (both the palette, and the radio button). 

If you look in previous PRs, there is a function `string_to_theme()` in /core/utils.rs which matches _ to Dark. I Think that should be Auto in this case?

**Comment by Frigyes06** [comment]
I tested this, it works

**Review by Frigyes06 — DISMISSED** [review]

**Comment by Frigyes06** [comment]
I think this will make auto the default. Is that what we want?

**Comment by Frigyes06** [comment]
Oh yeah, bottom of theme.rs...
What do we want for default? Auto, Dark, or Lupin?

**Comment by adhirajsinghchauhan** [comment]
> What do we want for default? Auto, Dark, or Lupin?
I vote for Auto, assuming it works on all OSes.

**Comment by Frigyes06** [comment]
Created a poll:
https://github.com/Universal-Debloater-Alliance/universal-android-debloater-next-generation/discussions/540

**Review by Frigyes06 — APPROVED** [review]

**Review by lavafroth — APPROVED** [review]
awesome!

**Comment by Rudxain** [comment]
> If you look in previous PRs, there is a function `string_to_theme()` in /core/utils.rs which matches _ to Dark.

Thanks! I forgot about that.

> I Think that should be Auto in this case?

When I first read that, I thought "No, don't do it! you'll mess up the code!", because I remember there's one part of the codebase where `Lupin` is default, and another where `Dark` is default. It was already a mess, so it would become worse 💀

> Default is lupin??

I was about to warn you! That's why we should use `default`, to keep everything consistent

File: src/core/utils.rs [hunk 2]
```diff
@@ -63,7 +63,7 @@ pub fn string_to_theme(theme: &str) -> Theme {
         "Dark" => Theme::Dark,
         "Light" => Theme::Light,
         "Lupin" => Theme::Lupin,
-        _ => Theme::Dark,
+        _ => Theme::Auto,
```
Comment (Rudxain) [inline src/core/utils.rs:66]: Should this be `default`?

File: src/core/theme.rs [hunk 3]
```diff
@@ -39,62 +46,73 @@ pub struct ColorPalette {
...
+            },
+        };
+        match self {
+            Self::Dark => DARK,
+            Self::Light => LIGHT,
+            Self::Lupin => LUPIN,
+            Self::Auto => match dark_light::detect() {
+                dark_light::Mode::Dark => DARK,
+                dark_light::Mode::Light => LIGHT,
+                // we should use `default` somehow,
```
Comment (Rudxain) [inline src/core/theme.rs:113]: If `Auto` itself is `default`, then I guess `Lupin` should stay hardcoded, unless the poll results in `Dark` beating `Lupin`

File: src/core/theme.rs [hunk 4]
```diff
@@ -39,62 +46,73 @@ pub struct ColorPalette {
 }
 
 impl Theme {
-    pub const ALL: [Self; 3] = [Self::Lupin, Self::Dark, Self::Light];
+    pub const ALL: [Self; 4] = [Self::Lupin, Self::Dark, Self::Light, Self::Auto];
```
Comment (Rudxain) [inline src/core/theme.rs:49]: If `Auto` is 0th variant, then it should be 0th here too

Comment (Frigyes06) [reply src/core/theme.rs:49]: It's your PR, feel free to commit.

Comment (Rudxain) [reply src/core/theme.rs]: I've read the message after the commit 😅

File: src/core/utils.rs (hunk 2 above)
Comment (Frigyes06) [reply src/core/utils.rs:66]: I *think* that works, let me try

Comment (Frigyes06) [reply src/core/utils.rs:66]: yep, that works!

**Comment by AnonymousWP** [comment]
This can be re-reviewed because of https://github.com/Universal-Debloater-Alliance/universal-android-debloater-next-generation/discussions/540.

File: src/core/theme.rs [hunk 5]
```diff
@@ -110,8 +110,7 @@ impl Theme {
             Self::Auto => match dark_light::detect() {
                 dark_light::Mode::Dark => DARK,
                 dark_light::Mode::Light => LIGHT,
-                // we should use `default` somehow,
-                // rather than hard-coding Lupin
+                // TO-DO: await #540
```
Comment (Frigyes06) [inline src/core/theme.rs]: This should not be Auto again, it should be one of the set themes. I think Dark would be most apptopriate

**Review by Frigyes06 — APPROVED** [review]

**Comment by Rudxain** [comment]
LGTM. This seems ready to merge 🎉

**Comment by Rudxain** [comment]
um... I noticed something weird:
image

Repro:

0. Open UAD from this branch/PR
1. Wait for device scan to stop (https://github.com/Universal-Debloater-Alliance/universal-android-debloater-next-generation/issues/532#issuecomment-2208799467)
2. Go to Settings
3. Ensure `Auto` theme is set
4. Change desktop theme
5. Click any empty space in the Settings View

It seems the view gets partially updated, for some reason. Switching to any other view, then going back to settings "fixes" it.

We should fix this later, as its priority doesn't seem high

**Comment by Frigyes06** [comment]
@Rudxain open a bug report lol

**Comment by Rudxain** [comment]
The PR was merged while I was writing it, lmao 💀. GH seriously needs something like Discord's "user is typing..."

**Comment by AnonymousWP** [comment]
But we should definitely fix this *before* v1.1.0.

**Pull request merged**